hack/test-go.sh

```diff
@@ -1,125 +1,223 @@
+            -o -path "${kubernetes_path}/pkg/api/v1"                      \
+            -o -path "${kubernetes_path}/test/e2e"                        \
+          \) -prune                                                       \
+        \) -name '*_test.go' -print0 | xargs -0n1 dirname -z | sort -uz | xargs -0n1 printf "${OS_GO_PACKAGE}/%s\n")"
+
+        test_packages="${test_packages} ${optional_kubernetes_packages}"
+    fi
+fi
+
+if [[ -n "${print_packages}" ]]; then
+    echo "The following packages will be tested by $0:"
+    for package in ${test_packages}; do
+        echo "${package}"
+    done
+    exit 0
 fi
 
-ret=$?; ENDTIME=$(date +%s); echo "$0 took $(($ENDTIME - $STARTTIME)) seconds"; exit "$ret"
+# Run 'go test' with the accumulated arguments and packages:
+if [[ -n "${junit_report}" ]]; then
+    # we need to generate jUnit xml
+    hack/build-go.sh tools/junitreport
+    junitreport="$(os::build::find-binary junitreport)"
+
+    if [[ -z "${junitreport}" ]]; then
+        echo "It looks as if you don't have a compiled junitreport binary"
+        echo
+        echo "If you are running from a clone of the git repo, please run"
+        echo "'./hack/build-go.sh tools/junitreport'."
+        exit 1
+    fi
+
+    test_output_file="${LOG_DIR}/test-go.log"
+    junit_report_file="${ARTIFACT_DIR}/report.xml"
+
+    # we don't care if 'go test' fails, we want to summarize the output anyway.
+    return_code=0 # default to 0, overwrite only if the following line fails
+    go test ${gotest_flags} ${test_packages} > ${test_output_file} 2>&1 || return_code=$?
+    echo "[INFO] Output from \`go test\` redirected to ${test_output_file}"
+
+    cat "${test_output_file}" | "${junitreport}" --type gotest --suites nested --roots github.com/openshift/origin > "${junit_report_file}"
+    echo "[INFO] jUnit XML report placed at ${junit_report_file}"
+    cat "${junit_report_file}" | "${junitreport}" summarize
+    exit "${return_code}"
+
+elif [[ -n "${coverage_output_dir}" ]]; then
+    # we need to generate coverage reports
+    for test_package in ${test_packages}; do
+        mkdir -p "${coverage_output_dir}/${test_package}"
+        local_gotest_flags="${gotest_flags} -coverprofile=${coverage_output_dir}/${test_package}/profile.out"
+
+        go test ${local_gotest_flags} ${test_package}
+    done
+
+    # assemble all profiles and generate a coverage report
+    echo 'mode: atomic' > "${coverage_output_dir}/profiles.out"
+    find "${coverage_output_dir}" -name profile.out | xargs sed '/^mode: atomic$/d' >> "${coverage_output_dir}/profiles.out"
+
+    go tool cover "-html=${coverage_output_dir}/profiles.out" -o "${coverage_output_dir}/coverage.html"
+    echo "[INFO] Coverage profile written to ${coverage_output_dir}/coverage.html"
+
+    # clean up all of the individual coverage reports as they have been subsumed into the report at ${coverage_output_dir}/coverage.html
+    rm -rf "${coverage_output_dir}/${OS_GO_PACKAGE}"
```

Comment:
I commented before that I found this cleanup confusing.  The comment you added helps, but I think the key point that I was missing before is that every path in `${test_packages}` has `${OS_GO_PACKAGE}` as a prefix (and thus every path `${coverage_output_dir}/${test_package}` has `${coverage_output_dir}/${OS_GO_PACKAGE}` as a prefix).  It might be worth adding a note to this effect to make the relationship between the `rm` command here and the `mkdir` and `go` commands above more blindingly obvious for folks like me who are a little slow.